packages/@aws-cdk/custom-resources/lib/provider-framework/provider.ts

```diff
@@ -0,0 +1,153 @@
+// tslint:disable: max-line-length
+import lambda = require('@aws-cdk/aws-lambda');
+import sfn = require('@aws-cdk/aws-stepfunctions');
+import tasks = require('@aws-cdk/aws-stepfunctions-tasks');
+import { Construct, Duration } from '@aws-cdk/core';
+import path = require('path');
+import consts = require('./runtime/consts');
+import { calculateRetryPolicy } from './util';
+
+const RUNTIME_HANDLER_PATH = path.join(__dirname, 'runtime');
+
+/**
+ * Initialization properties for the `Provider` construct.
+ */
+export interface ProviderProps {
+
+  /**
+   * The AWS Lambda function to invoke for all resource lifecycle operations
+   * (CREATE/UPDATE/DELETE).
+   *
+   * This function is responsible to begin the requested resource operation
+   * (CREATE/UPDATE/DELETE) and return any additional properties to add to the
+   * event, which will later be passed to `isComplete`. The `PhysicalResourceId`
+   * property must be included in the response.
+   */
+  readonly onEventHandler: lambda.IFunction;
+
+  /**
+   * The AWS Lambda function to invoke in order to determine if the operation is
+   * complete.
+   *
+   * This function will be called immediately after `onEvent` and then
+   * periodically based on the configured query interval as long as it returns
+   * `false`. If the function still returns `false` and the alloted timeout has
+   * passed, the operation will fail.
+   *
+   * @default - provider is synchronous. This means that the `onEvent` handler
+   * is expected to finish all lifecycle operations within the initial invocation.
+   */
+  readonly isCompleteHandler?: lambda.IFunction;
+
+  /**
+   * Time between calls to the `isComplete` handler which determines if the
+   * resource has been stabilized.
+   *
+   * The first `isComplete` will be called immediately after `handler` and then
+   * every `queryInterval` seconds, and until `timeout` has been reached or until
+   * `isComplete` returns `true`.
+   *
+   * @default Duration.seconds(5)
+   */
+  readonly queryInterval?: Duration;
+
+  /**
+   * Total timeout for the entire operation.
+   *
+   * The maximum timeout is 2 hours (yes, it can exceed the AWS Lambda 15 minutes)
+   *
+   * @default Duration.minutes(30)
+   */
+  readonly totalTimeout?: Duration;
+}
+
+/**
+ * Defines an AWS CloudFormation custom resource provider.
+ */
+export class Provider extends Construct {
+  /**
+   * The entrypoint of the custom resource provider.
+   *
+   * Use this when defining custom resources:
+   *
+   * @example
+   *
+   *    import cr = require('@aws-cdk/custom-resources');
+   *    import cfn = require('@aws-cdk/aws-cloudformation');
+   *
+   *    const provider = new cr.Provider(this, 'MyProvider', { ... })
+   *
+   *    new cfn.CustomResource(this, 'MyResource', {
+   *      provider: cfn.CustomResourceProvider.lambda(provider.entrypoint)
```

Comment:
Is it worth having this thing implement some kind of interface so we don't need to use `provider.entrypoint` ?

It could just be the provider.

Reply:
Yes, already changed this.